git-history.md

```diff
@@ -0,0 +1,69 @@
+# Git History Guidelines
+
+All of our projects follow a *fork + feature branch workflow*. That is, every
+project has a main *upstream* repository (e.g.
+[in-toto/in-toto](https://github.com/in-toto/in-toto)), with a *main* branch,
+called `master` or `develop` (e.g.
+[in-toto/in-toto@develop](https://github.com/in-toto/in-toto/tree/develop)).
+Contributors work on their own *fork* of the *upstream* repository (e.g.
+[lukpueh/in-toto](https://github.com/lukpueh/in-toto)). For every feature or
+fix they are working on, they create a *feature* branch, with a descriptive
+name (e.g.
+[lukpueh/in-toto@threshold-comment-fix](https://github.com/lukpueh/in-toto/tree/threshold-comment-fix)).
+A *feature* branch of a *fork* is incorporated into the *main* branch of the
+*upstream* repository by submitting a *pull request*.
+
+The rest of this document presents some guidelines to keep our projects' git
+histories clean.
+
+__*TL;DR*__
+Avoid merge commits in your *feature* branch. You can use `git rebase` to sync
+your *feature* branch with the *main* branch in *upstream* and also to keep
+your git history clean.
+
+
+__Choose the right base for your *feature* branch__
+In most cases you will base your *feature* branch on the *main* branch of your
+*fork*. Make sure that the *main* branch of your fork is in sync with the
+*main* branch of *upstream* so that you don't work on outdated code.
+Updating the *main* branch of your fork should not require *merge* commits,
+unless you have commited directly to your *main* branch, which you shouldn't!
+
+__Keep the base of your *feature* branch in sync with *upstream*__
+While you are working on your *feature* branch the *main* branch might move
+forward, because some other work was merged into the *main* branch. Check
+regularily, especially before you submit a *pull request*, if the
+base of your *feature* branch has moved forward. If it has, use `git rebase`
+to update the base and fix conflicts as they occur.
+
+__Tricks to keep the history of your *feature* branch clean (bonus task)__
+A clean history also means that you split your work into sensible chunks, i.e.
+commits (see [our commit guidelines](commits.md) for more details). You don't
+even need to worry about this while coding. You can use `git add` in
+`interactive` or `patch` mode to create multiple logically separated
+commits from a huge diff. Or use `git rebase` in `interactive` mode to rewrite
+the history of your *feature* branch to your hearts content.
+
+__Resolve conflicts in an open *pull request*__ Once you have submitted a *pull
+request* and you are receiving reviews and comments, you should avoid rebasing,
```

Comment:
While it's true that you should _hesitate_ before rebasing once the PR is up, there are still a lot of cases where it's still better to rebase than not:
- history cleanup / squashing
- removing unneeded accidental additions
- even IMHO conflict resolution: I think that conflict resolution is better with rebase than adding additional commits that won't make as much sense in the history once they're merged.  (BTW, would be interested if people disagree.  I do a lot of rebasing in PRs.)

Reply:
I totally agree with you on above items. There are two reasons why I think rebasing on PR's that are actively being reviewed should be avoided:
- AFAIK review code comments are bound to commit ids, hence break on rebase. I do have the feeling that GitHub made some enhancements in that regard, but I haven't checked.
- If I write a comment on someone's `Added foo` commit, that says "please modify bar", then it's easier for me to re-review, if the person pushes a `Modified bar in foo` commit, instead of e.g. silently modifying bar and force pushing it to `Added foo`.

However, once all the "please modify bar" comments are addressed and green-lighted by the reviewer, I think the contributor can/should squash the review-addressing commits, and use rebase to fix conflicts or clean up the history in general.

Reply:
> AFAIK review code comments are bound to commit ids, hence break on rebase.

Fancy a quick test?

Reply:
Will do

Reply:
As per the insights from my quick and fancy (ah ... dirty) [GitHub/rebase test below](https://github.com/secure-systems-lab/lab-guidelines/pull/14#issuecomment-440633959), I would conclude that GitHub does a pretty good job in tracking rebases. So you can probably disregard my first concern above. However, the second one is still relevant.

Reply:
I personally advice that in general we should keep git usage guidelines disconnected from github guidelines.